rotkehlchen/db/custom_assets.py

```diff
@@ -0,0 +1,154 @@
+        cursor.execute(query, bindings)
+        return [CustomAsset.deserialize_from_db(entry) for entry in cursor]
+
+    def get_custom_assets_and_limit_info(
+            self,
+            cursor: 'DBCursor',
+            filter_query: CustomAssetsFilterQuery,
+    ) -> Tuple[List[CustomAsset], int]:
+        """Returns a list of `CustomAsset` and a count of the assets that match the query.
+        May raise:
+        - InputError if swapped_from is not valid asset.
+        """
+        try:
+            custom_assets = self._get_custom_assets(
+                cursor=cursor,
+                filter_query=filter_query,
+            )
+            query, bindings = filter_query.prepare(with_pagination=False)
+            query = 'SELECT COUNT(*) from custom_assets AS A JOIN assets AS B ON A.identifier = B.identifier ' + query  # noqa: E501
+            total_found_result = cursor.execute(query, bindings).fetchone()[0]
+            return custom_assets, total_found_result
+        except UnknownAsset as e:
+            raise InputError(f'Unable to return custom assets due to: {str(e)}') from e
+
+    def add_custom_asset(self, custom_asset: CustomAsset) -> str:
+        """Adds a custom new custom asset into `assets` and `custom_assets` table.
+        May raise:
+        - InputError if the combination of the name & type / asset with identifier already exist.
+        """
+        result = self.check_if_custom_asset_exists(custom_asset)
+        if result is True:
+            raise InputError(
+                f'Custom asset with name "{custom_asset.name}" and type '
+                f'"{custom_asset.custom_asset_type}" already present in the database',
+            )
+
+        try:
+            with self.global_db.conn.write_ctx() as global_db_write_cursor, self.db.user_write() as db_write_cursor:  # noqa: E501
+                global_db_write_cursor.execute(
+                    'INSERT INTO assets(identifier, name, type, started, swapped_for) VALUES (?, ?, ?, ?, ?)',  # noqa: E501
+                    (
+                        custom_asset.identifier,
+                        custom_asset.name,
+                        AssetType.CUSTOM_ASSET.serialize_for_db(),
+                        custom_asset.started,
+                        custom_asset.swapped_for.identifier if custom_asset.swapped_for is not None else None,  # noqa: E501
+                    ),
+                )
+                global_db_write_cursor.execute(
+                    'INSERT INTO custom_assets(identifier, notes, type) VALUES(?, ?, ?)',
+                    custom_asset.serialize_for_db(),
+                )
+                self.db.add_asset_identifiers(db_write_cursor, [custom_asset.identifier])
+        except sqlite3.IntegrityError as e:
+            raise InputError(
+                f'Custom asset with name "{custom_asset.name}" and type '
```

Comment:
But I think what can happen here is that `swapped_for` is incorrect and the foreign key was hit